### DISCUSSION TITLE
Reading ZigBee Groups

### DISCUSSION THREAD

**Discussion by jamesonuk** [Koenkk/zigbee2mqtt#19133] [General]
I have just helped someone out with an issue that has caught me out before.

AIUI when you add a device to a group the link is actually stored on the device and which makes sense.

The issues come along when you either delete a group in Z2M front end or just delete a device that was in a group and add it back later.

In the front-end you only sree devices that have been assigned to the group in Z2M so you can end up with phantom devices that turn on / off with the group but don't show up as being in the group (which again makes sense as I presume the groups are stored as you add them.

Question is whether it is possible to do something a little slicker?   Is it possible to read the groups devices are assigned to and reflect this in Z2M?   I guess it would be difficult to make group removal any more robust as you will always have the scenario of unplugging a device whilst in a group and then not bring able to send a message to that device to remove it.

**Comment by jamesonuk** [comment]
I find myself coming back to this trying to debug an issue where I think a device might be linked to a group that I don't think it is.

Now looking at the docs I see there are some group membership commands against the group cluster.

command 0x02 says it is Get group membership with parameters groupcount and grouplist and states
> If the group count field of the command frame has a value of 0 indicating that the group list field is empty, the
entity SHALL respond with all group identifiers of which the entity is a member. 

and from the UI I can send command 0x02 to the group cluster (0x0004) with the payload
`{"groupcount: 0, grouplist: []}`
and I get 
 > Invoked '4.2' with payload '{"groupcount":0,"grouplist":[]}'

but I don't get a response which I guess means I am doing it wrong....

How can I see what comes back from this command?

**Comment by jamesonuk** [comment]
Bumping this again....    I believe I have a few bulbs which think they are in a group that Z2M does not.
The group data must be stored on the device so it knows whether or not to process group broadcast messages so presumably there is a way to read from it ???

**Reply by gregrs-uk** [reply]
You need to enable [debug logging](https://www.zigbee2mqtt.io/guide/configuration/logging.html#debugging) to see the message back. For example the following log extract:

```
debug 2025-01-03 13:39:48z2m:mqtt: Received MQTT message on 'zigbee2mqtt/Ensuite switch/1/set' with data '{"command":{"cluster":4,"command":2,"payload":{"groupcount":0,"grouplist":[]}}}'
debug 2025-01-03 13:39:48z2m: Publishing 'set' 'command' to 'Ensuite switch'
debug 2025-01-03 13:39:48z2m: Received Zigbee message from 'Ensuite switch', type 'commandGetMembershipRsp', cluster 'genGroups', data '{"capacity":32,"groupcount":0,"grouplist":[]}' from endpoint 1 with groupID 0
debug 2025-01-03 13:39:48z2m: No converter available for 'WS-EUK04' with cluster 'genGroups' and type 'commandGetMembershipRsp' and data '{"capacity":32,"groupcount":0,"grouplist":[]}'
info 2025-01-03 13:39:48zhc:tz: Invoked '4.2' with payload '{"groupcount":0,"grouplist":[]}'
```

**Reply by jamesonuk** [reply]
> You need to enable [debug logging](https://www.zigbee2mqtt.io/guide/configuration/logging.html#debugging) to see the message back. For example the following log extract:

I had done that but setting logs to debug in the UI doesn't seem to output the logs.   I have been through the docker logs and I can see the responses now.  Thanks  (Just with other command calls you see a response output in the standard pop ups).

Time to do some digging

**Reply by gregrs-uk** [reply]
Were you looking for responses in the Logs tab of the UI? If so, you need

```
advanced:
    log_debug_to_mqtt_frontend: true
```

in your `configuration.yaml`.

See docs [here](https://www.zigbee2mqtt.io/guide/configuration/logging.html#publishing-to-mqtt-and-to-frontend).

**Comment by jamesonuk** [comment]
Yeah I forgot about the logging changes a while back.   I only got as far
as being able to change the log level at runtime through the UI (which used
to work).

I was expecting a pop up (you only get the one saying a command was
submitted) but I will just use the Docker logs.

Now I just need to work out what is going on with the few odd lights....
